bin/mount-image-callback

```diff
@@ -118,14 +118,19 @@ do_umounts() {
 }
 
 cleanup() {
+	disconnect_qemu
 	if [ "${#UMOUNTS[@]}" -ne 0 ]; then
 		debug 2 "umounts: ${UMOUNTS[*]}"
-		do_umounts "${UMOUNTS[@]}"
+		do_umounts "${UMOUNTS[@]}" ||
+			{ error "failed cleaning up mounts"; return 1; }
```

Comment:
i may be wrong, but i think that UMOUNTS may *not* be all under TMPDIR.
if you told it where to mount (--mountpiont), then the mounts would be in a different location.

What i was suggsting was :+1: 
```bash
if [ -d "${TEMP_D}" ]; then
    umount_r "${TEMP_D}"
fi
```

the thing that we'd have to change is to make umount_r not fail if it did not unmount anything.

Reply:
Maybe not all UMOUNTS are under TMPDIR, but all the TMPDIR mounts (if any) should be in UMOUNTS, and this should be enough for us, I think?

Reply:
well you could have mounts that were done not by this program.

    mount-image-callback my.img --chroot -- mount -t /dev /dev  /dev

Reply:
I still don't get why we need an extra `umount_r`. TLDR: can you please provide a reproducer?

---

Assuming that you mean something like:
```
mount-image-callback my.img mchroot -- mount -t devtmpfs /dev /dev
```
from the script's point of view (i.e. from outside the chroot) the `/dev` mount will be under the mountpoint, which is under `TEMP_D` (if `--mountpoint` is not specified). The mountpoint is added to `UMOUNTS` and therefore umounted *recursively*. I verified this by doing
```
mount-image-callback autopkgtest-bionic-amd64.img mchroot -- sh -c 'mount -t devtmpfs /dev /dev ; sleep 60'
```
and in mounts I have:
```
/dev/nbd0p1 on /tmp/mount-image-callback.VQs86dyvlH/mp type ext4 (rw,relatime)
/dev on /tmp/mount-image-callback.VQs86dyvlH/mp/dev type devtmpfs (rw,relatime,size=12187540k,nr_inodes=3046885,mode=75mount_callback_umount5,inode64)
```
and everything is cleaned up correctly after `sleep` is done.

I also tried to manually mount something under the mountpoint from a different shell (mount not done by the script, not done from the chroot). It got unmounted just fine.

If I `cd mountpoint/dev` to interfere with the unmount process, I get:
```
$ mount-image-callback -vvv autopkgtest-images/autopkgtest-bionic-amd64.img mchroot -- sh -c 'mount -t devtmpfs /dev /dev ; sleep 30'
pid for /dev/nbd1 is 816180
connected /home/paride/ubuntu/autopkgtest-images/autopkgtest-bionic-amd64.img (qcow2) auto to /dev/nbd1. waiting for device.
partitioned disk.
mounted /dev/nbd1p1 via qemu-nbd /dev/nbd1 at /tmp/mount-image-callback.CMaV7qBW6m/mp
invoking: MOUNTPOINT=/tmp/mount-image-callback.CMaV7qBW6m/mp mchroot sh -c mount -t devtmpfs /dev /dev ; sleep 30
cmd returned 0. unmounting /tmp/mount-image-callback.CMaV7qBW6m/mp
umount: /tmp/mount-image-callback.CMaV7qBW6m/mp/dev: target is busy.
WARNING: unmounting filesystem at /tmp/mount-image-callback.CMaV7qBW6m/mp failed!
failed umount /home/paride/ubuntu/autopkgtest-images/autopkgtest-bionic-amd64.img
disconnecting /dev/nbd1
umounts: /tmp/mount-image-callback.CMaV7qBW6m/mp
umount: /tmp/mount-image-callback.CMaV7qBW6m/mp/dev: target is busy.
WARNING: unmounting filesystem at /tmp/mount-image-callback.CMaV7qBW6m/mp failed!
failed cleaning up mounts
```
as expected. Note that it tried twice: once from `mount_callback_umount()` and once from `cleanup()`.

If (e.g. because of a bug) the script silently failed to umount something, the `grep -q` will detect it and prevent the the `rm -Rf` when `TEMP_D` is in an unknown state.

If you still believe there's an issue here, could you please come up with a more concrete reproducer?

Also: do you think the current logic is in any significant way weaker than what we had before (`rm --one-file-system`)? If you don't think so, but still believe you spotted an issue which was already present, we could work on it outside of this "busybox compat" PR.

(Your idea of having `umount_r` return 2 in the "nothing to umount" case has some issues, but IMHO not worth discussing if we agree that we don't need it after all.)

Reply:
I think your right. 

I guess the one chance for failure is if umount_r falls. In that case,  then I guys there is no reason to rm at all. But check the umount return and avoid rm if it failed?

Reply:
The check is there:
```
do_umounts "${UMOUNTS[@]}" ||
    { error "failed cleaning up mounts"; return 1; }
```
If that `return 1` is reached the only cleanup operation that is skipped is the `rm -Rf $TEMP_D`.

Reply:
You’re right again. I was remembering how it was abs didn’t realize you’d changed that. 

So now I think you can drop the grep below on added line 128.

Reply:
We can, but I'm thinking of leaving it as an extra safety measure before the `rm -Rf`, just in case e.g. a future bug in `umount_r` makes it fail umount but not return failure, or UMOUNTS gets corrupted.

(Side note: TIL that util-linux umount has `--recursive`, but we're doing these changes for BusyBox compatibility, so that's not an option.)

Reply:
I added a comment about the `grep`, however I have no strong feelings about it, if you think it's useless or confusing I'll just drop it.

Reply:
if you have the grep to "protect", then youll never find out about the bug in umount_r.
I'm personally not a fan of extra code that "shouldnt ever be needed".
but I'll leave that up to you.

Reply:
Well if the grep fails an error is printed (= bugs in `umount_r` are found), and I was hoping that the latest version of the grep check was free from false positives (see my comment where I mention the fstab(5) format). For reference:

```
if ! grep -q "^[^ ]\+ ${TEMP_D}[/ ]" /proc/mounts; then
    rm -Rf "$TEMP_D" || error "removal of temp dir failed!"
else
    error "not removing temp dir (stale mounts)!"
fi
```

However I'm now convinced that the grep is unnecessary (it's especially your comment on "extra comment that shouldn't ever be needed" that made me reconsider). I now dropped it.